File: tests/unit/CommandFactory.test.js
const CommandFactory = require('../../server/src/commands/CommandFactory');

describe('CommandFactory', () => {
  describe('create', () => {
    it('returns MoveLeftCommand for MOVE_LEFT', () => {
      const cmd = CommandFactory.create('MOVE_LEFT');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('MOVE_LEFT');
    });
    it('returns MoveRightCommand for MOVE_RIGHT', () => {
      const cmd = CommandFactory.create('MOVE_RIGHT');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('MOVE_RIGHT');
    });
    it('returns StopCommand for STOP', () => {
      const cmd = CommandFactory.create('STOP');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('STOP');
    });
    it('returns JumpCommand for JUMP', () => {
      const cmd = CommandFactory.create('JUMP');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('JUMP');
    });
    it('returns AddScoreCommand for ADD_SCORE with amount arg', () => {
      const cmd = CommandFactory.create('ADD_SCORE', '10');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('ADD_SCORE');
    });
    it('returns SpawnItemCommand for SPAWN_ITEM with arg, db, bounds', () => {
      const bounds = { minX: -10, maxX: 10, minY: -10, maxY: 10 };
      const cmd = CommandFactory.create('SPAWN_ITEM', 'gem 0 0', null, bounds);
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('SPAWN_ITEM');
    });
    it('returns null for unknown type', () => {
      expect(CommandFactory.create('UNKNOWN')).toBeNull();
      expect(CommandFactory.create('')).toBeNull();
      expect(CommandFactory.create(null)).toBeNull();
    });
    it('normalizes type to uppercase', () => {
      const cmd = CommandFactory.create('move_left');
      expect(cmd).not.toBeNull();
      expect(cmd.type).toBe('MOVE_LEFT');
    });
  });
  describe('getValidTypes', () => {
    it('returns array including MOVE_LEFT, SPAWN_ITEM, ADD_SCORE', () => {
      const types = CommandFactory.getValidTypes();
      expect(Array.isArray(types)).toBe(true);
      expect(types).toContain('MOVE_LEFT');
      expect(types).toContain('ADD_SCORE amount');
      expect(types.some((t) => t.startsWith('SPAWN_ITEM'))).toBe(true);
    });
  });
});

File: socketrunner_unity/Assets/Tests/Unit/CommandFactoryTest.cs
using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;
using System.Collections;

public class CommandFactoryTest
{
	[Test]
	public void Create_NullData_ReturnsNull()
	{
		Assert.IsNull(CommandFactory.Create(null, null, null));
	}

	[Test]
	public void Create_ADD_SCORE_WithPayloadAmount_ReturnsAddScoreCommand()
	{
		var data = new NetworkData { type = "ADD_SCORE", payload = new PayloadData { amount = 5 } };
		ICommand cmd = CommandFactory.Create(data, null, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<AddScoreCommand>(cmd);
	}

	[Test]
	public void Create_ADD_SCORE_NoPayload_ReturnsAddScoreCommand()
	{
		var data = new NetworkData { type = "ADD_SCORE" };
		ICommand cmd = CommandFactory.Create(data, null, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<AddScoreCommand>(cmd);
	}

	[Test]
	public void Create_UnknownType_WithNullPlayer_ReturnsNull()
	{
		var data = new NetworkData { type = "MOVE_LEFT" };
		Assert.IsNull(CommandFactory.Create(data, null, null));
	}

	[UnityTest]
	public IEnumerator Create_MOVE_LEFT_WithPlayer_ReturnsMoveLeftCommand()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData { type = "MOVE_LEFT" };
		ICommand cmd = CommandFactory.Create(data, player, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<MoveLeftCommand>(cmd);

		Object.DestroyImmediate(go);
	}

	[UnityTest]
	public IEnumerator Create_MOVE_RIGHT_WithPlayer_ReturnsMoveRightCommand()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData { type = "MOVE_RIGHT" };
		ICommand cmd = CommandFactory.Create(data, player, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<MoveRightCommand>(cmd);

		Object.DestroyImmediate(go);
	}

	[UnityTest]
	public IEnumerator Create_STOP_WithPlayer_ReturnsStopCommand()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData { type = "STOP" };
		ICommand cmd = CommandFactory.Create(data, player, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<StopCommand>(cmd);

		Object.DestroyImmediate(go);
	}

	[UnityTest]
	public IEnumerator Create_JUMP_WithPlayer_ReturnsJumpCommand()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData { type = "JUMP" };
		ICommand cmd = CommandFactory.Create(data, player, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<JumpCommand>(cmd);

		Object.DestroyImmediate(go);
	}

	[UnityTest]
	public IEnumerator Create_SPAWN_ITEM_WithPayload_ReturnsSpawnItemCommand()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData
		{
			type = "SPAWN_ITEM",
			payload = new PayloadData { prefabType = "gem", x = 1f, y = 2f }
		};
		ICommand cmd = CommandFactory.Create(data, player, null);
		Assert.IsNotNull(cmd);
		Assert.IsInstanceOf<SpawnItemCommand>(cmd);

		Object.DestroyImmediate(go);
	}

	[UnityTest]
	public IEnumerator Create_SPAWN_ITEM_EmptyPrefabType_ReturnsNull()
	{
		var go = new GameObject();
		go.AddComponent<Rigidbody2D>();
		go.AddComponent<BoxCollider2D>();
		var player = go.AddComponent<PlayerController>();
		yield return null;

		var data = new NetworkData { type = "SPAWN_ITEM", payload = new PayloadData { prefabType = "" } };
		Assert.IsNull(CommandFactory.Create(data, player, null));

		Object.DestroyImmediate(go);
	}
}
